cmd/util/ledger/migrations/filter_unreferenced_slabs_migration.go

```diff
@@ -0,0 +1,146 @@
+package migrations
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/onflow/atree"
+	"github.com/onflow/cadence/runtime"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/cmd/util/ledger/reporters"
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/convert"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+func StorageIDFromRegisterID(registerID flow.RegisterID) atree.StorageID {
+	storageID := atree.StorageID{
+		Address: atree.Address([]byte(registerID.Owner)),
+	}
+	copy(storageID.Index[:], registerID.Key[1:])
+	return storageID
+}
+
+type FilterUnreferencedSlabsMigration struct {
+	log zerolog.Logger
+	rw  reporters.ReportWriter
+}
+
+var _ AccountBasedMigration = &FilterUnreferencedSlabsMigration{}
+
+const filterUnreferencedSlabsName = "filter-unreferenced-slabs"
+
+func NewFilterUnreferencedSlabsMigration(rwf reporters.ReportWriterFactory) *FilterUnreferencedSlabsMigration {
+	return &FilterUnreferencedSlabsMigration{
+		rw: rwf.ReportWriter(filterUnreferencedSlabsName),
+	}
+}
+
+func (m *FilterUnreferencedSlabsMigration) InitMigration(
+	log zerolog.Logger,
+	_ []*ledger.Payload,
+	_ int,
+) error {
+	m.log = log.
+		With().
+		Str("migration", filterUnreferencedSlabsName).
+		Logger()
+
+	return nil
+}
+
+func (m *FilterUnreferencedSlabsMigration) MigrateAccount(
+	_ context.Context,
+	address common.Address,
+	oldPayloads []*ledger.Payload,
+) (
+	newPayloads []*ledger.Payload,
+	err error,
+) {
+
+	migrationRuntime, err := NewAtreeRegisterMigratorRuntime(address, oldPayloads)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create migrator runtime: %w", err)
+	}
+
+	storage := migrationRuntime.Storage
+
+	newPayloads = oldPayloads
+
+	err = checkStorageHealth(address, storage, oldPayloads)
+	if err == nil {
+		return
+	}
+
+	// The storage health check failed.
+	// This can happen if there are unreferenced root slabs.
+	// In this case, we filter out the unreferenced root slabs from the payloads.
+
+	var unreferencedRootSlabsErr runtime.UnreferencedRootSlabsError
+	if !errors.As(err, &unreferencedRootSlabsErr) {
+		return nil, fmt.Errorf("storage health check failed: %w", err)
+	}
+
+	m.log.Warn().
+		Err(err).
+		Str("account", address.Hex()).
+		Msg("filtering unreferenced root slabs")
+
+	// Create a set of unreferenced root slabs.
+
+	unreferencedRootSlabIDs := map[atree.StorageID]struct{}{}
+	for _, storageID := range unreferencedRootSlabsErr.UnreferencedRootSlabIDs {
+		unreferencedRootSlabIDs[storageID] = struct{}{}
+	}
+
+	// Filter out unreferenced root slabs.
+
+	newCount := len(oldPayloads) - len(unreferencedRootSlabIDs)
+	newPayloads = make([]*ledger.Payload, 0, newCount)
+
+	filteredPayloads := make([]*ledger.Payload, 0, len(unreferencedRootSlabIDs))
```

Comment:
`unreferencedRootSlabIDs` only contains root slab IDs, so we need to filter out child slabs and referenced slabs as well.  

[PR 387](https://github.com/onflow/atree/pull/387) added a new function `(s *PersistentSlabStorage) getAllChildReferences(slab Slab)` to get all referenced slabs.  Maybe we can use it here when that function is exported.

Reply:
Thanks @fxamacker! Maybe we can also integrate https://github.com/onflow/atree/pull/387 into this PR/migration in general?

Reply:
@turbolent https://github.com/onflow/atree/pull/387 is merged, so you can use `GetAllChildReferences()` after you bump atree version.

Reply:
Given that this migration is for master, we'll actually need https://github.com/onflow/atree/pull/388, but no rush

Reply:
@turbolent These two PRs (for atree inlining branch) are ready for review:

- https://github.com/onflow/atree/pull/388 is updated with the same set of changes as https://github.com/onflow/atree/pull/387.  

- https://github.com/onflow/atree/pull/392 is opened (same as approved/merged PR https://github.com/onflow/atree/pull/391)

Reply:
Maybe I'm missing some context here.. Is it correct to think that https://github.com/onflow/atree/pull/387 and this are two different problems? i.e: "references to non-existent registers" vs "unreferenced registers"?

Reply:
@SupunS Yes, they are two separate problems. The work in e.g. https://github.com/onflow/atree/pull/387 to handle one of the problems (cleaning up missing slabs), depends on/provides a function (getting all child references) that is also needed for the other problem here (cleaning up unreferenced slabs)

Reply:
@turbolent Should we target this PR to a different branch?  `master` branch doesn't have updated atree and it might be easier if we target this to Cadence migration branch for example and I can cherry-pick it to Atree migration branch.

Reply:
@turbolent There are some breaking changes if the target is [feature/atree-inlining-cadence-v0.42](https://github.com/onflow/flow-go/tree/feature/atree-inlining-cadence-v0.42).  

Do you want to maybe target `feature/stable-cadence` first, so we don't need to change it twice due to breaking changes?  

I can port this PR to atree inlining feature branches after it is merged.

Reply:
It started out targeting `feature/stable-cadence`, but then I rebased it onto `master`.

I now rebased it back, and will continue to update it (e.g. use `GetAllChildReferences`)

Reply:
@fxamacker Added the call to `GetAllChildReferences` in [c2e87fd](https://github.com/onflow/flow-go/pull/5653/commits/c2e87fd0285eec48a696f9a90273bf81a5866b90). As far as I understand it is already recursing down and returns all the child slabs of the child slabs and so forth, correct?

Reply:
> Added the call to GetAllChildReferences in [c2e87fd](https://github.com/onflow/flow-go/pull/5653/commits/c2e87fd0285eec48a696f9a90273bf81a5866b90). As far as I understand it is already recursing down and returns all the child slabs of the child slabs and so forth, correct?

@turbolent Yes, also nested containers in separate slabs, etc.